src/transform/desugar_destructuring.rs

```diff
@@ -17,17 +17,67 @@
 //! let d = x.b in
 //! ...
 //! ```
+//!
+//! ## The function pattern
+//! ```text
+//! let f = fun x @ {a, b=c}, {d ? 2, ..w} => <do_something> in ...
+//! ```
+//! will be transformed to:
+//! ```text
+//! let f = fun x %unnamed% => (
+//!     let {a, b=c} = x in
+//!     let {d ? 2, ..w} = %unnamed% in
+//!     <do_something>
+//! ) in ...
+//! ```
 use crate::destruct::{Destruct, Match};
 use crate::identifier::Ident;
 use crate::term::make::{op1, op2};
 use crate::term::{
     BinaryOp::DynRemove, BindingType, MetaValue, RichTerm, Term, UnaryOp::StaticAccess,
 };
 
-/// Wrap the desugar term in a meta value containing the "Record contract" needed to check the
+/// Entry point of the patterns desugaring.
+/// It call:
+/// - `desugar_with_contract` when `rt` is a let pattern.
+/// - `desugar_fun` when `rt` is a function with patterns as arguments (`Term::FunPattern`).
+pub fn transform_one(rt: RichTerm) -> RichTerm {
+    match *rt.term {
+        Term::LetPattern(..) => desugar_with_contract(rt),
+        Term::FunPattern(..) => desugar_fun(rt),
+        _ => rt,
+    }
+}
+
+/// Desugar a function with patterns as arguments.
+/// This function does not perform nested transformation because internaly it's only used in a top
+/// down traversal. It mean that the return value is a normal `Term::Fun` but it can contains
+/// `Term::FunPattern` and `Term::LetPattern` inside RHS.
+pub fn desugar_fun(rt: RichTerm) -> RichTerm {
+    match *rt.term {
+        Term::FunPattern(x, pat, t_) if !pat.is_empty() => {
+            let x = x.unwrap_or_else(super::fresh_var);
+            RichTerm::new(
+                Term::Fun(
+                    x.clone(),
+                    Term::LetPattern(None, pat, Term::Var(x).into(), t_).into(),
```

Comment:
I'm not sure. because, exept for the actual body of the function, for a multiparams function, the pos is same as the main function. So should we not preferably put the `rt.pos` than the `t_.pos`?